File: AEDs/Algoritmos de Ordenacao/C/Quick.c
#include "Geracao.h"
#include "Log.h"

void quick ( int *arranjo, int esq, int dir, Log* log ) 
{
    int i = esq, j = dir;
    int pivo = arranjo[(dir+esq)/2];
    incrementarMovimentacoes( log, 1 );
    
    incrementarComparacoes( log, 1);
    while( i <= j ) 
    {
        incrementarComparacoes( log, 1 );
        while( arranjo[i] < pivo )
        {
            incrementarComparacoes( log, 1 );
            i++;
        } // end while

        incrementarComparacoes( log, 1 );
        while( arranjo[j] > pivo ) 
        {
            incrementarComparacoes( log, 1 );
            j--;
        } // end while

        incrementarComparacoes( log, 1 );
        if( i <= j ) 
        {
            swap( arranjo + i, arranjo + j );
            incrementarMovimentacoes( log, 3 );
            i++;
            j--;
        } // end if
    } // end while

    incrementarComparacoes( log, 1 );
    if( esq < j ) {
        quick( arranjo, esq, j, log );
    } // end if

    incrementarComparacoes( log, 1 );
    if( i < dir ) {
        quick( arranjo, i, dir, log );
    } // end if
} // end quick ( )

void sort ( int* arranjo, int tamanho, Log* log ) {
    quick( arranjo, 0, tamanho-1, log );
} // end sort ( )

int main ( void )
{
    Geracao* arranjo = new_Geracao( 100 );
    Log*     log     = new_Log ( );

    printf( "\n//----- QUICK SORT -----//\n");

    preencher_aleatorio( arranjo, 200 );

    printf("\nARRANJO PRE ORDENACAO:\n");
    mostrar( arranjo );

    inicioTempo( log );
    sort( arranjo->arranjo, arranjo->tamanho, log );
    fimTempo( log );

    printf("\nARRANJO POS ORDENACAO:\n");
    mostrar( arranjo );

    registrar ( log );

    delete_Log ( log );
    delete_Geracao( arranjo );
} // end main ( )

File: Codigos/AEDs/Algoritmos de Ordenacao/Java/Quick.java
public class Quick extends Geracao 
{
    public Quick ( )
    {
        super ( );
    }

    public Quick ( int tamanho )
    {
        super( tamanho );
    }

    /**
     *  <p>Quick Sort é um algoritmo de ordenação eficiente que divide o arranjo em
     *  partições, selecionando um elemento como pivô, e rearranjando os elementos
     *  de modo que os menores que o pivô fiquem à sua esquerda e os maiores à sua
     *  direita.<p>
     *  
     *  <p>Complexidade: O(n*log(n)) no caso médio e melhor caso, O(n^2) no pior caso.            </p>
     *  <p>Número de comparações: O(n*log(n)) no caso médio e melhor caso, O(n^2) no pior caso.   </p> 
     *  <p>Número de movimentações: O(n*log(n)) no caso médio e melhor caso, O(n^2) no pior caso. </p>
     * 
     *  @param left : primeira posição do arranjo.
     *  @param right : ultima posição do arranjo.
     */
    public void quickSort ( int left, int right, Log log ) 
    {
        int i = left, j = right;
        int pivot = arranjo[(left + right)/2];
        log.incrementarMovimentacoes();

        log.incrementarComparacoes();
        while( i <= j ) 
        {
            log.incrementarComparacoes();
            while( arranjo[i] < pivot ) {
                i++;
                log.incrementarComparacoes();
            } // end while

            log.incrementarComparacoes();
            while( arranjo[j] > pivot ) {
                j--;
                log.incrementarComparacoes();
            } // end while

            log.incrementarComparacoes();
            if( i <= j ) {
                swap( i, j );
                log.incrementarMovimentacoes(3);
                i++;
                j--;
            } // end if
        } // end while

        log.incrementarComparacoes();
        if( left < j ) {
            quickSort( left, j, log );
        } // end if
        
        log.incrementarComparacoes();
        if( i < right ) {
            quickSort( i, right, log );
        } // end if
    } // end quickSort ( )

    @Override
    public void sort ( Log log )
    {
        quickSort( 0, tamanho-1, log );
    } // end sort ( )

    public static void main( String[] args ) 
    {
        Quick arranjo = new Quick( 100 );
        Log log = new Log();

        System.out.println( "\n//----- QUICK SORT -----//");

        arranjo.preencher_aleatorio();

        System.out.println("\nARRANJO PRE ORDENACAO:");
        arranjo.mostrar( );

        log.inicioTempo();
        arranjo.sort( log );
        log.fimTempo();

        System.out.println("\nARRANJO POS ORDENACAO:");
        arranjo.mostrar( );

        log.mostrar();
    } // end main ( )
}
